### ISSUE TITLE
[Performer] Biased estimate of orthogonal matrix in Performer code.

### ISSUE THREAD

**Issue by Parskatt** [google-research/google-research#436]
At line https://github.com/google-research/google-research/blob/master/performer/fast_self_attention/fast_self_attention.py#L357 the r matrix from the q,r decomposition is ignored. This is not good, since the r matrix is needed to ensure that the samples are uniform over the sphere. See for example the init in pytorch: https://pytorch.org/docs/stable/_modules/torch/nn/init.html#orthogonal_

and the paper https://arxiv.org/pdf/math-ph/0609050.pdf

Not performing this fix significantly increases the variance of the estimate.

**Comment by kchoro** [comment]
Hi Parskatt, thank you for the comment. So you are right that for unbiased sampling from the Haar measure, flipping of the columns via R should be done, however here ultimately we want unbiased low-variance estimation of the kernel value, not of the matrix. If flipping is not done, the marginal distributions  of rows (after chi(d)-renormalization) are still Gaussian so kernel estimator remains unbiased. Furthermore, since all the proofs effectively rely only on the fact that marginal distributions of rows match gaussian and different rows are exactly orthogonal, all the results regarding lower variance for the orthogonal case remain valid regardless of whether you use R or not. So ultimately whether you use R or not should not matter when we talk about the accuracy of kernel estimation.

**Comment by Parskatt** [comment]
Hi, thanks for the response.

I'm not quite sure I follow your reasoning.

I think the distributions are not actually still gaussian after renormalization if you use the biased estimation.

I'm sorry if I'm completely misunderstanding your argument.  I provide a notebook below to show what I mean with biased estimation:

https://colab.research.google.com/drive/1xMSh8c8E8cDK9GqBpVEcGPKHDV6W44C-?usp=sharing

What I get in the special case that d=2 is that literally no rows have an angle between [-pi/2, 0]. I don't see how renormalization would solve this problem. It also means that estimates of any k,q lying in that range would be an underestimate, which increases the variance of the estimate.

**Comment by Parskatt** [comment]
The simlest case (which is slightly ridicoulus), is `q,r = torch.qr(torch.ones(1,1))` and `q,r = torch.qr(-torch.ones(1,1))`. These always produce the same q.

**Comment by kchoro** [comment]
Hi Parskatt,

My point is that you can get low MSE kernel estimation (*strictly more accurate than iid*) without having exactly Haar measure in the implementation. This is what is done in many papers on ORFs (for reference: https://papers.nips.cc/paper/6246-orthogonal-random-features). Lets conclude this thread with the statement that if you want to have exact Haar measure, you should definitely flip according to R and it is definitely not a bad idea (thank you for catching this !). I would encourage you to run Performers with R-based flipping and without on a couple of downstream tasks and if you see additional gains in downstream tasks from flipping, please let us know right away :)

**Comment by Parskatt** [comment]
Thanks for the response. 

I agree with you that in higher dimensions this probably doesnt affect the MSE as much.

From what I've observed it does increase variance in all cases, but that doesn't necessarily lead to better performance on downstream tasks.

I'll close this issue for now :)
